src/vanilla/utils/atomWithObservable.ts

```diff
@@ -81,107 +74,86 @@ export function atomWithObservable<Data>(
-    const start = () => {
-      if (subscription) {
-        clearTimeout(timer)
-        subscription.unsubscribe()
-      }
+    const startSubscription = () => {
+      subscription?.unsubscribe()
       subscription = observable.subscribe({
-        next: (d) => listener({ d }),
-        error: (e) => listener({ e }),
+        next: (value) => {
+          resolve(value)
+          preMountData = value
+          setResult?.(value)
+        },
+        error: (error) => {
+          reject(error)
+          if (error instanceof Error) {
+            // @ts-expect-error TypeScript doesn't allow setting atom to error
+            // If we use Promise.reject here instead, Vitest will complain about an uncaught Error
+            setResult?.(error)
+          } else {
+            setResult?.(Promise.reject(error))
+          }
+        },
         complete: () => {},
       })
-      if (isNotMounted() && options?.unstable_timeout) {
-        timer = setTimeout(() => {
-          if (subscription) {
-            subscription.unsubscribe()
-            subscription = undefined
-          }
-        }, options.unstable_timeout)
-      }
     }
-    start()
-
-    const resultAtom = atom(lastResult || initialResult)
 
     if (import.meta.env?.MODE !== 'production') {
       resultAtom.debugPrivate = true
     }
 
-    resultAtom.onMount = (update) => {
-      setResult = update
-      if (lastResult) {
-        update(lastResult)
+    startSubscription()
+
+    resultAtom.onMount = (setAtom) => {
+      setResult = setAtom
+
+      if (preMountData !== EMPTY) {
+        setAtom(preMountData)
       }
-      if (subscription) {
-        clearTimeout(timer)
-      } else {
-        start()
+
+      if (subscription === undefined) {
+        startSubscription()
       }
+
       return () => {
-        setResult = undefined
-        if (subscription) {
-          subscription.unsubscribe()
-          subscription = undefined
-        }
+        subscription?.unsubscribe()
+        subscription = undefined
       }
     }
-    return [resultAtom, observable, makePending, start, isNotMounted] as const
-  })
 
-  if (import.meta.env?.MODE !== 'production') {
-    observableResultAtom.debugPrivate = true
-  }
+    return { atom: resultAtom, observable }
+  })
 
-  const observableAtom = atom(
-    (get) => {
-      const [resultAtom] = get(observableResultAtom)
-      const result = get(resultAtom)
-      if (result instanceof Promise) {
-        return result.then(returnResultData)
-      }
-      return returnResultData(result)
-    },
-    (get, set, data: Data) => {
-      const [resultAtom, observable, makePending, start, isNotMounted] =
-        get(observableResultAtom)
+  return atom(
+    (get) => get(get(observableResultAtom).atom),
+    (get, _, data: Data) => {
+      const observable = get(observableResultAtom).observable
       if ('next' in observable) {
-        if (isNotMounted()) {
-          set(resultAtom, makePending())
```

Comment:
I wonder how it's resolved this case.

Reply:
I don't believe we need to re-trigger the promise here if it has already been resolved?